newIDE/app/src/CommandPalette/CommandManager.js

```diff
@@ -0,0 +1,103 @@
+// @flow
+import { type MessageDescriptor } from '../Utils/i18n/MessageDescriptor.flow';
+
+type CommandHandler = () => void | Promise<void>;
+
+export type Command = {|
+  displayText: MessageDescriptor,
+  enabled: boolean,
+  handler: CommandHandler,
+|};
+
+export type NamedCommand = {|
+  name: string,
+  ...Command,
+|};
+
+export type CompoundCommandOption<T> = {|
+  value: T,
+  handler: CommandHandler,
+  text: string,
+  iconSrc?: string,
+|};
+
+export type CompoundCommand<T> = {|
+  displayText: MessageDescriptor,
+  enabled: boolean,
+  generateOptions: () => Array<CompoundCommandOption<T>>,
+|};
+
+export type NamedCompoundCommand<T> = {|
+  name: string,
+  ...CompoundCommand<T>,
+|};
+
+export default class CommandManager {
+  commands: { [string]: Command };
+  compoundCommands: { [string]: CompoundCommand<any> };
+  _isScoped: boolean;
+
+  constructor(scoped: ?boolean) {
+    this.commands = {};
+    this.compoundCommands = {};
+    this._isScoped = scoped || false;
+  }
+
+  registerCommand = (commandName: string, command: Command) => {
+    if (this.commands[commandName])
+      return console.warn(
+        `Tried to register command ${commandName}, but it is already registered.`
+      );
+    this.commands[commandName] = command;
+  };
+
+  deregisterCommand = (commandName: string) => {
+    if (!this.commands[commandName])
+      return console.warn(
+        `Tried to deregister command ${commandName}, but it is not registered.`
+      );
+    delete this.commands[commandName];
+  };
+
+  registerCompoundCommand = <T>(
```

Comment:
In the same way, I think we should see if we can have the commands registered in the same way, be it a command or a command with options.
At least, commands are indexed by names, and I see no reason for allowing a command and a command with options that have the same name to be allowed. Unfortunately, currently as we're storing them in separate objects, this situation can arise.

Let's see if we can store them in a single object?

Reply:
Done.